### PR TITLE
Use Blaze 3.8 (exclusively), remove PointerVector

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

As we had hoped, the recent updates in Blaze 3.8 allow us to get rid of our PointerVector wrapper altogether, trimming a great deal of fairly complicated code from the repository.

Of course, this is a breaking change because in order to make these simplifications, we have to demand the use of 3.8.

In several places the utilities that use the blaze `CustomVector` now use the operator `*` rather than `~`, as Blaze has deprecated `~`, and now overloads `*` to have the same functionality.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: docs/DevGuide/ImplementingVectors.md [hunk 1]
```diff
@@ -45,13 +45,29 @@ DataVector>`. This template pattern is known as the
...
-will be implicitly inherited from `VectorImpl`. The assignment and constructors
-may be inherited calling the following alias code in the vector class
-definition:
+operators, and iterator generation members.
+Most of these are inherited from Blaze types, but in addition, the methods
+`set_data_ref`, and `pup` are defined for use in SpECTRE.
+All except for the assignment operators and constructors will be implicitly
+inherited from `VectorImpl`.
+The assignment and constructors may be inherited calling the following alias
+code in the vector class definition:
```
Comment (wthrowe) [inline docs/DevGuide/ImplementingVectors.md]: You've done some weird formatting adding lots of newlines in this file, and it's caused strange formatting in the rendered version in some places (the last code bit in the last sentence on the page, for example).

File: src/DataStructures/VectorImpl.hpp [hunk 2]
```diff
@@ -281,7 +295,7 @@ VectorImpl<T, VectorType>& VectorImpl<T, VectorType>::operator=(
   if (this != &rhs) {
     if (owning_) {
       owned_data_ = std::move(rhs.owned_data_);
-      ~*this = ~rhs; /* PointerVector is trivially copyable */
+     **this = std::move(*rhs);
```
Comment (wthrowe) [inline src/DataStructures/VectorImpl.hpp]: Formatting.

File: src/Utilities/Blaze.hpp [hunk 3]
```diff
@@ -41,4 +40,76 @@ constexpr bool useDefaultInitialization = false;
 
 // Disable HPX parallelization
 #define BLAZE_HPX_PARALLEL_MODE 0
+
+// Disable all padding
+#define BLAZE_USE_PADDING 0
```
Comment (wthrowe) [inline src/Utilities/Blaze.hpp:43]: Does this mean Blaze will never add padding, or Blaze assumes that nothing is padded?  I'm not sure the latter is a reasonable constraint.

File: src/Utilities/Blaze.hpp [hunk 4]
```diff
@@ -41,4 +40,76 @@ constexpr bool useDefaultInitialization = false;
...
+  BLAZE_ALWAYS_INLINE decltype(auto) load(const T& a) const noexcept {
+    BLAZE_CONSTRAINT_MUST_BE_SIMD_PACK(T);
+    return step_function(a);
+  }
+};
+}  // namespace blaze
+
+template <typename VT, bool TF>
+// NOLINTNEXTLINE(readability-const-return-type)
+BLAZE_ALWAYS_INLINE decltype(auto) step_function(
```
Comment (wthrowe) [inline src/Utilities/Blaze.hpp:103]: As far as I can tell, this function and the next are identical except for name, so why does this one need a NOLINT and the other does not?

File: tests/Unit/Evolution/Systems/CurvedScalarWave/Test_UpwindFlux.cpp [hunk 5]
```diff
@@ -33,6 +33,8 @@
 #include "Utilities/MakeWithValue.hpp"
 #include "Utilities/ProtocolHelpers.hpp"
 
+#include "Utilities/TmplDebugging.hpp"
+
```
Comment (wthrowe) [inline tests/Unit/Evolution/Systems/CurvedScalarWave/Test_UpwindFlux.cpp]: Remove.

**Review by wthrowe — CHANGES_REQUESTED** [review]
I'm not really qualified to review the details of the Blaze integration changes, but what I understood looked fine, and that may be the best tested code in SpECTRE so I'm satisfied with it.

File: src/Utilities/Blaze.hpp (hunk 3 above)
Comment (moxcodes) [reply src/Utilities/Blaze.hpp:43]: This flag means Blaze won't add padding.

File: src/Utilities/VectorAlgebra.hpp [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #pragma once
 
 #include "Utilities/Gsl.hpp"
-#include "Utilities/PointerVector.hpp"  // For blaze::MultTrait
+#include "Utilities/Blaze.hpp"
```
Comment (wthrowe) [inline src/Utilities/VectorAlgebra.hpp:6]: Alphabetize

File: src/Utilities/Blaze.hpp [hunk 7]
```diff
@@ -4,6 +4,11 @@
 /// \file
 /// Includes Blaze library with specific configs
 
+#include <stdio.h>
```
Comment (wthrowe) [inline src/Utilities/Blaze.hpp]: I don't think you need this, but if you do use `<cstdio>`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
A couple of trivial things I missed before.  Otherwise looks good.  Squash directly.

**Review by wthrowe — DISMISSED** [review]

**Comment by moxcodes** [comment]
Oops sorry, looks like something broke that I overlooked in `Variables` following the rebase, fixing now...

**Comment by moxcodes** [comment]
force-pushed the fix that checks that the variables size is not zero before resetting the underlying blaze vector, otherwise blaze errors because it refuses to assign a `nullptr` on a reset.
Cheers!

**Review by wthrowe — DISMISSED** [review]

File: support/Environments/frontera_gcc.sh [hunk 8]
```diff
@@ -64,7 +64,7 @@ EOF
         echo "Installing Blaze..."
```
Comment (nilsdeppe) [inline support/Environments/frontera_gcc.sh:64]: I've added Blaze 3.8 to Wheeler as a module. Could you update the two Wheeler envs and verify that works?

**Review by nilsdeppe — COMMENTED** [review]
LGTM! Just need to update Wheeler env (I've added the module)

**Comment by nilsdeppe** [comment]
Once the Wheeler env has been verified working I'll update the container and then we can merge this :)

File: support/Environments/frontera_gcc.sh (hunk 8 above)
Comment (moxcodes) [reply support/Environments/frontera_gcc.sh:64]: Thanks! I'll check that and confirm here once I've made sure everything builds and tests alright.

**Comment by moxcodes** [comment]
Sorry for the delay -- I've confirmed that it builds and tests on wheeler with the updated env in the new commit.
Cheers!

File: docs/DevGuide/ImplementingVectors.md [hunk 9]
```diff
@@ -45,14 +45,26 @@ DataVector>`. This template pattern is known as the
...
+```
 The class template `VectorImpl` defines various constructors, assignment
 operators, and iterator generation members. Most of these are inherited from
 Blaze types, but in addition, the methods `set_data_ref`, and `pup` are defined
 for use in SpECTRE. All except for the assignment operators and constructors
 will be implicitly inherited from `VectorImpl`. The assignment and constructors
 may be inherited calling the following alias code in the vector class
-definition:
-
+ definition:
```
Comment (kidder) [inline docs/DevGuide/ImplementingVectors.md]: why the space before `definition`?

**Review by kidder — COMMENTED** [review]
you can squash

**Review by fmahebert — COMMENTED** [review]
Looking at the output of `git grep PointerVector`, I think a few comments still reference this class and should be updated, and a few scripts in `tools` have special handling of the PointerVector file that is no longer needed.

**Comment by moxcodes** [comment]
fixup for the residual appearances of `PointerVector` posted.
Cheers!

**Comment by fmahebert** [comment]
You can squash the fixup 👍

**Comment by moxcodes** [comment]
Fixup squashed in.
Cheers!

**Review by wthrowe — DISMISSED** [review]

**Comment by moxcodes** [comment]
rebased, squashed in a missing include, and added a commit for the unrelated fixes from the rebase.
Hopefully, this will pass CI now.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by nilsdeppe** [comment]
We can fix the clang-tidy error in a separate PR. Here is what it is:
```
/__w/spectre/spectre/src/NumericalAlgorithms/LinearSolver/Lapack.cpp:85:13: error: Value stored to 'output_vector_size' during its initialization is never read [clang-analyzer-deadcode.DeadStores,-warnings-as-errors]
Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
  const int output_vector_size = matrix_operator->columns();
1 warning treated as error
            ^
```
This is in release mode. The variable is (I think) only used in a `ASSERT` statement.

**Pull request merged**